src/lib/utils/MathExtra.ts

```diff
@@ -327,6 +328,69 @@ const inner = (a, b) => {
   throw new Error(`Cannot process with the invalid inputs ${a} and ${b}`);
 };
 
+/**
+ * Converts 1D matrix of size N to a 2D diagonal matrix of size NxN
+ * @example makeDiagonal([1, 2, 3])
+ * returns [[1, 0, 0], [0, 2, 0], [0, 0, 3]
+ * @param x - source array
+ */
+const makeDiagonal = (X: Type1DMatrix<number>): Type2DMatrix<number> => {
+  const xSize = X.length;
+  if (xSize === 0) {
+    throw new Error("Can't make diagonal matrix from an empty array");
+  }
+
+  const result = [];
+  for (let i = 0; i < xSize; ++i) {
+    const curArray = [];
+    for (let j = 0; j < xSize; ++j) {
+      curArray.push(j === i ? X[i] : 0);
+    }
+    result.push(curArray);
+  }
+
+  return result;
+};
+
+/**
+ * Computes sigmoid function in point x
+ * @param x - point in which sigmoid is computed
+ */
+const sigmoid = (x: number): number => {
+  if (x >= 0) {
+    const z = Math.exp(-x);
+    return 1 / (1 + z);
+  } else {
+    const z = Math.exp(x);
+    return z / (1 + z);
```

Comment:
why can't we just put all these in one line? 

`Math.exp(x) / (1 + z)`

Reply:
Because that way we save one operation.

Reply:
ah